### ISSUE TITLE
Connection problems with 6.0.2 (Broken pipe, Redis server went away...)

### ISSUE THREAD

**Issue by potsky** [phpredis/phpredis#2437]
### Expected behaviour

We would like to use version 6.0.2 as we used the 5.3.7 without connection problems.

### Actual behaviour

After about 10 hours of operation, we start getting these kind of errors when hundreds of writes are done in a `foreach` loop on a single backend server:
- Redis server went away
- Redis::exec(): Send of 2504 bytes failed with errno=32 Broken pipe
- Redis::hMset(): Send of 1431 bytes failed with errno=32 Broken pipe
- ...

We run about 30k jobs by day with [Laravel Horizon](https://laravel.com/docs/10.x/horizon), so we had hundreds of thousands exceptions with 6.0.2 in just a few days.

### Found workaround

1. Restarting our web servers resolves temporarily the problem.
2. [We rolled back](https://github.com/Scalingo/php-buildpack/issues/395) on Scalingo PaaS to **phpredis 5.3.7 and everything is fine**.

### Current investigations

- There is no error on the Redis server logs
- We have checked the count of connections from our backend to Redis cluster and we stay under 100, including cluster connections, which is a far cry from the default 10k connections.
- The Redis servers metrics are ok, there is no memory leak, no CPU peak, ...
- This is not a network problem in the datacenter given that it works fine with 5.3.7

### I'm seeing this behaviour on

- **Infrastructure:** Scalingo PaaS
- **OS:** Linux app-sirenergies-one-off-8185 5.4.0-121-generic # 137-Ubuntu SMP Wed Jun 15 13:33:07 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- **Redis:** 7.2.3 (Image [7.2.3-1](https://hub.docker.com/r/scalingo/redis))
- **PHP:** 8.2.14
- **phpredis:** 6.0.2
- **Cluster configuration:**
<img width="414" alt="Screenshot 2024-01-21 - 13-17-23@2x" src="https://github.com/phpredis/phpredis/assets/408237/501e6d5d-ca00-4349-ae41-1f3fb76e6ebd">

### Steps to reproduce, backtrace or example script

I have no idea on how to reproduce this behaviour because it's not systematic. For example, when we send SMS to our customers at 8am and reboot the servers 1h before at 7am, we don't have any problems. On the other hand, when we only rebooted them the day before at 8pm, they crash at 8am the next day, even though it's the same code, the same volume, ...

**phpinfo on 5.3.7**

<img width="958" alt="5 3 7" src="https://github.com/phpredis/phpredis/assets/408237/ee0e2258-b16f-4d09-94ff-ee490fca2e0b">

**phpinfo on 6.0.2**

<img width="967" alt="6 0 2" src="https://github.com/phpredis/phpredis/assets/408237/e5ab0424-3a07-4c45-8bf2-6a842464a067">

- The Redis sentinel version hase changed from 0.1 to 1.0
- `redis.session.early_refresh` is new on 6.0.2
- The default value of `redis.session.lock_retries` has changed from `10` to `100`
- The default value of `redis.session.lock_wait_time` has changed from `2000` to `20000`

### I've checked

- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

**Comment by michael-grunder** [comment]
The broken pipe (errno 32) error often happens when sending huge payloads, such as massive multi-exec or pipeline blocks, which hit kernel limits.

However, your situation appears slightly different.  The trick is going to be reproducing the behavior.

Laravel horizon are long-running job queues, right?  Maybe we can simulate similar activity with runners that execute the same Redis commands and same general volume?  Perhaps PhpRedis 6.0.2 has a bug where we continue to try and use a socket even after it's failed.

Another option would be to run one of your jobs under something like `rr`.  If you replicate the problem that way it would almost certainly identify exactly what's going wrong.  The debugging would need to be on your side though because it would record everything including all of the payloads to and from Redis.

**Comment by potsky** [comment]
HI @michael-grunder,

yep Laravel Horizon is a long-running job queue tool like Sidekiq for example. The possible problem with a dead socket is an idea but how to test this? Complicated if think.

Using `rr` is a very good idea. We need to check this with the Scalingo team:
- what kind ov VM do they use because it seems to me that rr did not work on all VM hosts
- how to get debug logs because given that we run on ephemeral servers, the local disk is not accessible

To be continued...

**Comment by gbalduzzi** [comment]
I'm seeing the very same issue on the same setup: Laravel horizon with a lot of jobs, occasional Broken Pipe error.

Did you manage to find something?

I also tried to debug the problem by connecting locally to the cloud redis istance (on digital ocean) but it happening only after hours and with a lot of traffic make it very hard to reproduce

**Comment by rechl** [comment]
@gbalduzzi We had the exact same issue on Digital Ocean with Laravel Horizon. Our best guess was:
Different PHP processes on the same machine share once TCP connection to redis. If the connection fails due to network errors, other processes still tried to use this stale connection. Simply disabling persistent connections for Redis helped. 
But I would not recommend to disable persistent connections completely, as this will slow down you application a lot. And _never_ disable it for client facing code...(other story)

We now are setting a persistent ID for each PHP Process (remember, these processes get started by by the Horizon daemon / supervisor). To achieve it we did this in the `boot()` section of the AppServiceProvider:


```php
// when running in horizon or any console statement use a separate connection for each process
if (app()->runningInConsole()) {
    config([
        'database.redis.queue.persistent_id' => 'redis-pid-'.getmypid().'-queue',
    ]);
}
```

And this in the `database.php`

```php
'queue' => [
    'url' => env('REDIS_URL'),
    'host' => env('REDIS_HOST'),
    'password' => env('REDIS_PASSWORD')),
    'port' => env('REDIS_PORT'),
    'database' => env('REDIS_DB'),
    'timeout' => 30,
    // be aware: if the app is running in console we set a persistent_id in AppServiceProvider::boot() for each process
    'persistent' => true
],

```

**This removed 99% of the errors.** 

A upgrade to the latest PHP version, latest Redis version and ext-redis to the latest version helped us. I can confirm the code above in combination with `ext-redis: 6.1.0` works.
Also we analyzed the `6.1.0` source code and could not find any bugs there regarding persistent connections.(We did not yet try the `6.2.0` version)

In April 2024 we upgraded from `5.3.7` to `6.0.2`. That was a really bad decision and withing 20 minutes we had to roll back.  

### **Do not use `ext-redis 6.0.0 up to 6.0.2`**. 

We are still seeing Redis errors with `6.1.0` , but much less and now it is mainly `Redis::exec(): SSL: Broken pipe`  

Beside the "broken" `6.0.x` ext-redis release:

## TLDR:  We are pretty confident there is an TCP connection handling error in either PHP, openssl (😱) or Linux but not in ext-redis...

... which is unfortunately way beyond our capabilities.😢

If anyone finds the real bug, we definitely owe more then coffee! Feel free to drop me a line. 


Please tell me if this worked for you. 😃

**Comment by gbalduzzi** [comment]
@rechl You're a lifesaver! 🙌 

I've updated phpredis and added the persistent ID connection. Initial results look good.
I'll continue to monitor things for a few days and will let you know how it goes. 

Thanks again!

**Comment by michael-grunder** [comment]
Closing as this seems to have been resolved.

**Comment by rechl** [comment]
@michael-grunder we are still actively investigating the issue. I will update once we found a real solution.

**Comment by jarnovanleeuwen** [comment]
> We had the exact same issue on Digital Ocean with Laravel Horizon. Our best guess was:
> Different PHP processes on the same machine share once TCP connection to redis.

We solved it on DigitalOcean's Managed Valkey by increasing the Idle Connection Timeout from 300 seconds (DigitalOcean's default) to unlimited (Valkey's default).
